File: templates/backends/python-fastapi/app/repository/user_repo.py
from __future__ import annotations

from datetime import datetime, timezone

from sqlalchemy import delete, func, or_, select, update
from sqlalchemy.ext.asyncio import AsyncSession

from app.domain.errors import ERR_NOT_FOUND, DomainError
from app.domain.login_event import LoginEvent
from app.domain.token import EmailVerificationToken, PasswordResetToken, RefreshToken
from app.domain.twofa import RecoveryCode, TwoFactorToken
from app.domain.user import User
from app.dto.admin import UserListParams

ALLOWED_SORT_COLUMNS = {"id", "first_name", "last_name", "email", "role", "language", "created_at", "updated_at"}


class UserRepository:
    def __init__(self, session: AsyncSession):
        self._session = session

    async def find_by_id(self, user_id: int) -> User:
        stmt = select(User).where(User.id == user_id, User.deleted_at.is_(None))
        result = await self._session.execute(stmt)
        user = result.scalar_one_or_none()
        if user is None:
            raise DomainError(ERR_NOT_FOUND, "User not found")
        return user

    async def find_by_public_id(self, public_id: str) -> User:
        stmt = select(User).where(User.public_id == public_id, User.deleted_at.is_(None))
        result = await self._session.execute(stmt)
        user = result.scalar_one_or_none()
        if user is None:
            raise DomainError(ERR_NOT_FOUND, "User not found")
        return user

    async def find_by_email(self, email: str) -> User:
        stmt = select(User).where(func.lower(User.email) == email.lower(), User.deleted_at.is_(None))
        result = await self._session.execute(stmt)
        user = result.scalar_one_or_none()
        if user is None:
            raise DomainError(ERR_NOT_FOUND, "User not found")
        return user

    async def find_by_email_unscoped(self, email: str) -> User | None:
        stmt = select(User).where(func.lower(User.email) == email.lower())
        result = await self._session.execute(stmt)
        return result.scalar_one_or_none()

    async def find_by_username(self, username: str) -> User:
        stmt = select(User).where(func.lower(User.username) == username.lower(), User.deleted_at.is_(None))
        result = await self._session.execute(stmt)
        user = result.scalar_one_or_none()
        if user is None:
            raise DomainError(ERR_NOT_FOUND, "User not found")
        return user

    async def find_by_email_or_username(self, identifier: str) -> User:
        stmt = select(User).where(
            or_(
                func.lower(User.email) == identifier.lower(),
                func.lower(User.username) == identifier.lower(),
            ),
            User.deleted_at.is_(None),
        )
        result = await self._session.execute(stmt)
        user = result.scalar_one_or_none()
        if user is None:
            raise DomainError(ERR_NOT_FOUND, "User not found")
        return user

    async def is_username_available(self, username: str, exclude_id: int | None = None) -> bool:
        stmt = select(User.id).where(func.lower(User.username) == username.lower(), User.deleted_at.is_(None))
        if exclude_id:
            stmt = stmt.where(User.id != exclude_id)
        result = await self._session.execute(stmt)
        return result.scalar_one_or_none() is None

    async def create(self, user: User) -> None:
        self._session.add(user)
        await self._session.flush()

    async def update(self, user: User) -> None:
        await self._session.flush()

    async def soft_delete(self, user_id: int) -> None:
        await self._session.execute(
            update(User).where(User.id == user_id).values(deleted_at=datetime.now(timezone.utc))
        )
        await self._session.flush()

    async def hard_delete(self, user_id: int) -> None:
        await self._session.execute(delete(User).where(User.id == user_id))
        await self._session.flush()

    async def hard_delete_related_records(self, user_id: int) -> None:
        for model in (EmailVerificationToken, TwoFactorToken, RefreshToken, RecoveryCode, PasswordResetToken, LoginEvent):
            await self._session.execute(delete(model).where(model.user_id == user_id))
        await self._session.flush()

    async def list_users(self, params: UserListParams) -> tuple[list[User], int]:
        base = select(User).where(User.deleted_at.is_(None))
        count_stmt = select(func.count(User.id)).where(User.deleted_at.is_(None))

        if params.search:
            pattern = f"%{params.search}%"
            search_filter = or_(
                User.first_name.ilike(pattern),
                User.last_name.ilike(pattern),
                User.email.ilike(pattern),
                func.concat(User.first_name, " ", User.last_name).ilike(pattern),
            )
            base = base.where(search_filter)
            count_stmt = count_stmt.where(search_filter)

        total_result = await self._session.execute(count_stmt)
        total = total_result.scalar() or 0

        sort_col = params.sort_field if params.sort_field in ALLOWED_SORT_COLUMNS else "id"
        col = getattr(User, sort_col)
        order = col.desc() if params.sort_order.upper() == "DESC" else col.asc()
        offset = (params.page - 1) * params.limit

        stmt = base.order_by(order).offset(offset).limit(params.limit)
        result = await self._session.execute(stmt)
        users = list(result.scalars().all())
        return users, total

    async def count(self) -> int:
        r = await self._session.execute(select(func.count(User.id)).where(User.deleted_at.is_(None)))
        return r.scalar() or 0

    async def count_verified(self) -> int:
        r = await self._session.execute(
            select(func.count(User.id)).where(User.deleted_at.is_(None), User.verified.is_(True))
        )
        return r.scalar() or 0

    async def count_by_role(self, role: str) -> int:
        r = await self._session.execute(
            select(func.count(User.id)).where(User.deleted_at.is_(None), User.role == role)
        )
        return r.scalar() or 0

    async def count_since(self, since: datetime) -> int:
        r = await self._session.execute(
            select(func.count(User.id)).where(User.deleted_at.is_(None), User.created_at >= since)
        )
        return r.scalar() or 0

File: backend/go-api/internal/repository/user_repo.go
package repository

import (
	"context"
	"errors"
	"fmt"
	"time"

	"saas-starter/backend/go-api/internal/domain"
	"saas-starter/backend/go-api/internal/dto"
	"saas-starter/backend/go-api/internal/platform/database"

	"gorm.io/gorm"
)

type userRepo struct {
	db *gorm.DB
}

func NewUserRepository(db *gorm.DB) UserRepository {
	return &userRepo{db: db}
}

func (r *userRepo) conn(ctx context.Context) *gorm.DB {
	return database.DBFromContext(ctx, r.db)
}

func (r *userRepo) FindByID(ctx context.Context, id uint) (*domain.User, error) {
	var user domain.User
	if err := r.conn(ctx).First(&user, id).Error; err != nil {
		if errors.Is(err, gorm.ErrRecordNotFound) {
			return nil, domain.ErrNotFound
		}
		return nil, err
	}
	return &user, nil
}

func (r *userRepo) FindByPublicID(ctx context.Context, publicID string) (*domain.User, error) {
	var user domain.User
	if err := r.conn(ctx).Where("public_id = ?", publicID).First(&user).Error; err != nil {
		if errors.Is(err, gorm.ErrRecordNotFound) {
			return nil, domain.ErrNotFound
		}
		return nil, err
	}
	return &user, nil
}

func (r *userRepo) FindByEmail(ctx context.Context, email string) (*domain.User, error) {
	var user domain.User
	if err := r.conn(ctx).Where("LOWER(email) = LOWER(?)", email).First(&user).Error; err != nil {
		if errors.Is(err, gorm.ErrRecordNotFound) {
			return nil, domain.ErrNotFound
		}
		return nil, err
	}
	return &user, nil
}

func (r *userRepo) FindByEmailUnscoped(ctx context.Context, email string) (*domain.User, error) {
	var user domain.User
	if err := r.conn(ctx).Unscoped().Where("email = ?", email).First(&user).Error; err != nil {
		if errors.Is(err, gorm.ErrRecordNotFound) {
			return nil, domain.ErrNotFound
		}
		return nil, err
	}
	return &user, nil
}

func (r *userRepo) FindByUsername(ctx context.Context, username string) (*domain.User, error) {
	var user domain.User
	if err := r.conn(ctx).Where("LOWER(username) = LOWER(?)", username).First(&user).Error; err != nil {
		if errors.Is(err, gorm.ErrRecordNotFound) {
			return nil, domain.ErrNotFound
		}
		return nil, err
	}
	return &user, nil
}

func (r *userRepo) Create(ctx context.Context, user *domain.User) error {
	return r.conn(ctx).Create(user).Error
}

func (r *userRepo) Update(ctx context.Context, user *domain.User) error {
	return r.conn(ctx).Save(user).Error
}

func (r *userRepo) SoftDelete(ctx context.Context, id uint) error {
	return r.conn(ctx).Delete(&domain.User{}, id).Error
}

func (r *userRepo) HardDelete(ctx context.Context, id uint) error {
	return r.conn(ctx).Unscoped().Delete(&domain.User{}, id).Error
}

func (r *userRepo) HardDeleteRelatedRecords(ctx context.Context, userID uint) error {
	db := r.conn(ctx)
	tables := []interface{}{
		&domain.EmailVerificationToken{},
		&domain.TwoFactorToken{},
		&domain.RefreshToken{},
		&domain.RecoveryCode{},
		&domain.PasswordResetToken{},
		&domain.LoginEvent{},
	}
	for _, model := range tables {
		if err := db.Where("user_id = ?", userID).Delete(model).Error; err != nil {
			return fmt.Errorf("failed to delete %T for user %d: %w", model, userID, err)
		}
	}
	return nil
}

func (r *userRepo) IsUsernameAvailable(ctx context.Context, username string, excludeID ...uint) (bool, error) {
	query := r.conn(ctx).Where("LOWER(username) = LOWER(?)", username)
	if len(excludeID) > 0 && excludeID[0] > 0 {
		query = query.Where("id != ?", excludeID[0])
	}
	var user domain.User
	if err := query.First(&user).Error; err != nil {
		if errors.Is(err, gorm.ErrRecordNotFound) {
			return true, nil
		}
		return false, err
	}
	return false, nil
}

var allowedSortColumns = map[string]string{
	"id": "id", "first_name": "first_name", "last_name": "last_name",
	"email": "email", "role": "role", "language": "language",
	"created_at": "created_at", "updated_at": "updated_at",
}

func (r *userRepo) List(ctx context.Context, params dto.UserListParams) ([]domain.User, int64, error) {
	var users []domain.User
	var total int64

	query := r.conn(ctx).Model(&domain.User{})

	if params.SearchQuery != "" {
		pattern := "%" + params.SearchQuery + "%"
		query = query.Where(
			"first_name ILIKE ? OR last_name ILIKE ? OR email ILIKE ? OR (first_name || ' ' || last_name) ILIKE ?",
			pattern, pattern, pattern, pattern,
		)
	}

	if err := query.Count(&total).Error; err != nil {
		return nil, 0, err
	}

	sortCol := "id"
	if col, ok := allowedSortColumns[params.SortField]; ok {
		sortCol = col
	}
	sortDir := "ASC"
	if params.SortOrder == "desc" || params.SortOrder == "DESC" {
		sortDir = "DESC"
	}
	offset := (params.Page - 1) * params.Limit

	if err := query.Order(sortCol + " " + sortDir).
		Limit(params.Limit).Offset(offset).
		Find(&users).Error; err != nil {
		return nil, 0, err
	}

	return users, total, nil
}

func (r *userRepo) CountByRole(ctx context.Context, role string) (int64, error) {
	var count int64
	err := r.conn(ctx).Model(&domain.User{}).Where("role = ?", role).Count(&count).Error
	return count, err
}

func (r *userRepo) Count(ctx context.Context) (int64, error) {
	var count int64
	err := r.conn(ctx).Model(&domain.User{}).Count(&count).Error
	return count, err
}

func (r *userRepo) CountVerified(ctx context.Context) (int64, error) {
	var count int64
	err := r.conn(ctx).Model(&domain.User{}).Where("verified = ?", true).Count(&count).Error
	return count, err
}

func (r *userRepo) CountSince(ctx context.Context, since time.Time) (int64, error) {
	var count int64
	err := r.conn(ctx).Model(&domain.User{}).Where("created_at >= ?", since).Count(&count).Error
	return count, err
}
